Beams/Logic/vtkSlicerMlcPositionLogic.cxx

```diff
@@ -0,0 +1,1082 @@
+
+        double side2 = 0.0;
+        if (FindLeafAndTargetCollision( beamNode, leaf2->GetOutputPort(), beamInverseTransformFilter->GetOutputPort(),
+          side2, InitialPos2 + 1., 2, typeMLCX))
+        {
+          table->SetValue( leafPair, 2, side2);
+        }
+        else
+        {
+          vtkWarningMacro("CalculateMultiLeafCollimatorPosition: Side 2 collision hasn't been found");
+        }
+
+    } // leafPair for cycle ends
+
+  }
+
+  return true;
+}
+
+//---------------------------------------------------------------------------
+bool
+vtkSlicerMlcPositionLogic::FindLeafAndTargetCollision( vtkMRMLRTBeamNode* vtkNotUsed(beamNode), 
+  vtkAlgorithmOutput* leafOutput, vtkAlgorithmOutput* targetOutput,
+  double& sidePos, double initialPosition, int sideType, bool mlcType, 
+  double maxPositionDistance, double positionStep)
+{
+  int contactMode = 1;
+  auto targetMatrix = vtkSmartPointer<vtkMatrix4x4>::New(); // fixed target matrix
+  targetMatrix->Identity();
+  auto leafTransform = vtkSmartPointer<vtkTransform>::New(); // moving leaf transform
+  leafTransform->Identity();
+  auto collide = vtkSmartPointer<vtkCollisionDetectionFilter>::New();
+  collide->SetInputConnection( 0, leafOutput);
+  collide->SetTransform( 0, leafTransform);
+  collide->SetInputConnection( 1, targetOutput);
+  collide->SetMatrix( 1, targetMatrix);
+  collide->SetBoxTolerance(0.0);
+  collide->SetCellTolerance(0.0);
+  collide->SetNumberOfCellsPerNode(2);
+  if (contactMode == 0)
+  {
+    collide->SetCollisionModeToAllContacts();
+  }
+  else if (contactMode == 1)
+  {
+    collide->SetCollisionModeToFirstContact();
+  }
+  else
+  {
+    collide->SetCollisionModeToHalfContacts();
+  }
+  collide->GenerateScalarsOn();
+
+  // Move the leaf
+  int numSteps = fabs(initialPosition) + fabs(maxPositionDistance) / fabs(positionStep);
+  if (sideType == 1)
+  {
+      if (mlcType) // MLCX
+      {
+        leafTransform->Translate( initialPosition, 0.0, 0.0);
+      }
+      else // MLCY
+      {
+        leafTransform->Translate( 0., initialPosition, 0.0);
+      }
+  }
+  else
+  {
+      if (mlcType) // MLCX
+      {
+        leafTransform->Translate( initialPosition, 0.0, 0.0);
+      }
+      else // MLCY
+      {
+        leafTransform->Translate( 0., initialPosition, 0.0);
+      }
+  }
+
+  bool res = false;
+  double p[3] = {};
+  for (int i = 0; i < numSteps; ++i)
+  {
+    if (sideType == 1)
+    {
+      if (mlcType) // MLCX
+      {
+        leafTransform->Translate( positionStep, 0.0, 0.0);
+      }
+      else // MLCY
+      {
+        leafTransform->Translate( 0.0, positionStep, 0.0);
+      }
+    }
+    else
+    {
+      if (mlcType) // MLCX
+      {
+        leafTransform->Translate( -1. * positionStep, 0.0, 0.0);
+      }
+      else // MLCY
+      {
+        leafTransform->Translate( 0.0, -1. * positionStep, 0.0);
+      }
+    }
+    
+    leafTransform->Update();
+    collide->Update();
+
+    if (collide->GetNumberOfContacts() > 0)
```

Comment:
There are practically no comments in your code. For the sake of maintainability, can you please go through the whole class and add simple sentences for the major things it does? For example here, without spending several minutes I would have no idea what we need to do if there is collision.

Reply:
More detailed description has been added.